File: packages/sdk/src/resources/fhir.ts
import type { HttpClient } from '../http';
import type { FhirSearchResponse } from '../types';

export class FhirClient {
  constructor(private readonly http: HttpClient) {}

  search(resourceType: string, params: Record<string, string | number> = {}): Promise<FhirSearchResponse> {
    const q = new URLSearchParams(Object.entries(params).map(([k, v]) => [k, String(v)]));
    const qs = q.toString() ? `?${q}` : '';
    return this.http.get<FhirSearchResponse>(`/fhir/${resourceType}${qs}`);
  }

  read(resourceType: string, id: string): Promise<Record<string, unknown>> {
    return this.http.get<Record<string, unknown>>(`/fhir/${resourceType}/${id}`);
  }

  create(resource: Record<string, unknown>): Promise<Record<string, unknown>> {
    const rt = resource['resourceType'] as string;
    return this.http.post<Record<string, unknown>>(`/fhir/${rt}`, resource);
  }

  update(resource: Record<string, unknown>): Promise<Record<string, unknown>> {
    const rt = resource['resourceType'] as string;
    const id = resource['id'] as string;
    return this.http.put<Record<string, unknown>>(`/fhir/${rt}/${id}`, resource);
  }

  transaction(bundle: Record<string, unknown>): Promise<Record<string, unknown>> {
    return this.http.post<Record<string, unknown>>('/fhir/Bundle', bundle);
  }
}

File: sdk/python/pe/resources/fhir.py
from typing import Any
import httpx


class FhirClient:
    def __init__(self, http: httpx.Client):
        self._http = http

    def search(self, resource_type: str, **params: Any) -> dict:
        return self._http.get(f"/fhir/{resource_type}", params=params).json()

    def read(self, resource_type: str, id: str) -> dict:
        return self._http.get(f"/fhir/{resource_type}/{id}").json()

    def create(self, resource: dict) -> dict:
        rt = resource["resourceType"]
        return self._http.post(f"/fhir/{rt}", json=resource).json()

    def update(self, resource: dict) -> dict:
        rt = resource["resourceType"]
        rid = resource["id"]
        return self._http.put(f"/fhir/{rt}/{rid}", json=resource).json()

    def transaction(self, bundle: dict) -> dict:
        return self._http.post("/fhir/Bundle", json=bundle).json()
